corehq/apps/geospatial/models.py

```diff
@@ -117,3 +117,48 @@ def allocate_objectives(self):
                 gdf.drop(closest_user_id, inplace=True)
 
         return user_assignment
+
+    def group_objectives(self):
+        """
+        Group objectives that are closest together, within the constraints of `max_distance` and `max_assignable`.
+        Returns a 2D list containing all the objectives groups.
+        [['obj_a', 'obj_b']]
+
+        At least one of the constraints has to be non-null, otherwise will simply return the list of objective ids.
+        """
+        if not self.objectives:
+            return []
+        if not self.max_distance and not self.max_assignable:
+            return [[obj.id for obj in self.objectives]]
+
+        self.objectives.sort(key=lambda x: (x.lat, x.lon))
```

Comment:
Do we care about sorting?

I'm not sure it helps. If ...

    >>> objectives = [(1.01, 1.01), (1.02, 1.02), (1.01, 10), (1.02, 10)]
    >>> sorted(objectives, key=lambda x: (x[0], x[1]))
    [(1.01, 1.01), (1.01, 10), (1.02, 1.02), (1.02, 10)]

I haven't tested this, but I think that if `max_assignable` is 2, then the first bin will  be `[(1.01, 1.01), (1.01, 10)]` and the second bin will be `[(1.02, 1.02), (1.02, 10)]`?

Reply:
Since we are going through the objectives list in a sequential fashion, the sorting is done to essentially sort the relative distance of the objectives from each other. Doing it like this allows us to more easily group objectives together with a similar location into the same bin. Given the sequential nature of the algorithm, if sorting is not applied, then we might get objective x that is close to objective y but will be put into a separate bin if its index is earlier/later in the list.

Using the example list you gave, assuming that it is in the format (lat, lon), the function gives the following bins: `[[(1.01, 1.01), (1.01, 10)], [(1.02, 1.02), (1.02, 10)]]` which is correct when comparing their distances from each other using the `great_circle()` helper function. If sorting had not been applied, then they would have been grouped as `[[(1.01, 1.01), (1.02, 1.02)], [(1.01, 10), (1.02, 10)]]` which would be incorrect when comparing their relative distances.

For interest's sake, you can also test the validity of the function using some real-world test data (taken from South Africa, grouped by province) as follows:
```
objectives = [
    Objective('sandton', lat=-26.106768103016243, lon=28.055971657040097), # Gauteng
    Objective('nelspruit_central', lat=-25.46985876996438, lon=30.98046450782144), # Mpumalanga
    Objective('midrand', lat=-25.99976379136922, lon=28.124031758870697), # Gauteng
    Objective('white_river', lat=-25.330952842423393, lon=31.015423703762927), # Mpumalanga
    Objective('west_acres', lat=-25.490537104958616, lon=30.95593614354665), # Mpumalanga
]
oa = ObjectiveAllocator(None, objectives, 200, 2)
bins = oa.group_objectives()
```
The function should return the following bins: `[['sandton', 'midrand'], ['west_acres', 'nelspruit_central', 'white_river']]`.

Reply:
Sorry @zandre-eng I really didn't explain that very well.

> sorting is done to essentially sort the relative distance of the objectives from each other.

What I meant was that the `sort()` method here is sorting by latitude first, and after that by longitude. As a result, locations with a similar latitude are sorted close to each other, even if there is a big difference in longitude.

In the example you gave, White River, Nelspruit and West Acres are sorted to the front, followed by Midrand and Sandton. But what if we take a more extreme example. If we don't specify a `max_distance` ...

```python
In [1]: from corehq.apps.geospatial.models import Objective, ObjectiveAllocator

In [2]: objectives = [
   ...:     Objective('windhoek', lat=-22.57, lon=17.083611),  # Windhoek
   ...:     Objective('capetown', lat=-33.925278, lon=18.423889),  # Cape Town
   ...:     Objective('antofagasta', lat=-23.65, lon=-70.4),  # Antofagasta, Chile
   ...:     Objective('santiago', lat=-33.45, lon=-70.666667),  # Santiago, Chile
   ...: ]

In [3]: oa = ObjectiveAllocator(None, objectives, None, 2)

In [4]: bins = oa.group_objectives()

In [5]: bins
Out[5]: [['capetown', 'santiago'], ['antofagasta', 'windhoek']]
```
... then `sort()` will put Cape Town in the same bin as Santiago, and Windhoek with Antofagasta, because their latitudes are close, even though their longitudes are far.

Reply:
I think we can take two possible approaches here:
1. Use a library, as Sravan suggested. I think this is the best approach.
2. Solve this, maybe just because it's fun. I did a little digging, and realized that the answer is found in linear algebra; stuff I covered in second-year math, which for me was a very, very long time ago. If you're curious, and this is as hazy for you as it is for me, you're looking for [linear algebra](https://ocw.mit.edu/courses/18-06sc-linear-algebra-fall-2011/), and [elementary graph theory](https://ocw.mit.edu/courses/6-042j-mathematics-for-computer-science-fall-2010/).

Reply:
Thanks for the clarification @kaapstorm! I agree that using a library would be the best solution, for reasons stated in [this](https://github.com/dimagi/commcare-hq/pull/33153#issuecomment-1604046935) comment. If we were to resolve this issue however, then another idea that pops to mind (asides from the one you suggested) is to use geopandas dataframes. We can implement a loop that fetches an "anchor" objective, and then uses a dataframe to get all objectives within the given constraints. This set of objectives can be removed from selection and put into a bin, and then we iterate again until all objectives have been grouped.